experimental/PlaneCurve/EllCurveZnZ.jl

```diff
@@ -0,0 +1,714 @@
+# Miller Rabin Test
+################################################################################
+function _Miller_Rabin_witness(N::fmpz, a::fmpz, even::fmpz, odd::fmpz)
+    R = ResidueRing(ZZ, N)
+    x = ZZ(R(a)^odd)
+    if x == ZZ(1) || x == N - ZZ(1)
+        return false
+    else
+        for i = 1:nbits(even)
+            x = ZZ(R(x)^2)
+            if x == N - ZZ(1)
+                return false
+            end
+        end
+    end
+    return true
+end
+
+################################################################################
+
+@doc Markdown.doc"""
+    Miller_Rabin_test(N::fmpz, k::Int64 = 20)
+
+Given an odd number `N`, return `false` if the number is composite, and `true` if it
+is probably prime.
+"""
+function Miller_Rabin_test(N::fmpz, k::Int64 = 20)
+    !iszero(N % 2) || error("an odd number is expected")
+    (even, odd) = evenodd(N - ZZ(1))
+    for i = 1:k
+        a = rand(ZZ(2):min(N, ZZ(10^6)))
+        if _Miller_Rabin_witness(N, a, even, odd)
+            return false
+        end
+    end
+    return true
+end
+
+################################################################################
+# Pollard's methods
+################################################################################
+@doc Markdown.doc"""
+    Pollard_rho(N::fmpz, bound::Int = 50000)
+
+The algorithm computes a factor of `N` and returns it.
+"""
+function Pollard_rho(N::fmpz, bound::Int = 50000)
+    R = ResidueRing(ZZ, N)
+    x = rand(ZZ(2):N-1)
+    y = rand(ZZ(2):N-1)
+    d = ZZ(1)
+    i = 1
+    while d == 1 && i <= bound
+        x = ZZ(R(x)^2 + R(1))
+        y = R(y)^2 + R(1)
+        y = y + R(1)
+        d = gcd(x - ZZ(y), N)
+        i = i + 1
+    end
+    return d
+end
+
+################################################################################
+
+@doc Markdown.doc"""
+    Pollard_p_1(N::fmpz, B::fmpz = ZZ(10)^5)
+
+The algorithm computes a factor of `N` and returns it.
+"""
+function Pollard_p_1(N::fmpz, B::fmpz = ZZ(10)^5)
+    p = [i for i in PrimesSet(ZZ(1), B)]
+    x = rand(ZZ(2):N-1)
+    y = x
+    c = ZZ(0)
+    i = 0
+    j = i
+    k = length(p)
+    w = 0
+
+    while w == 0
+        i = i + 1
+        if i > k
+            g = gcd(x - ZZ(1), N)
+            if g == ZZ(1)
+                return g
+            else
+                i = j
+                x = y
+                w = 1
+            end
+        else
+            q = p[i]
+            qq = q
+            l = div(B, q)
+            while qq <= ZZ(l)
+                qq = q * qq
+            end
+            x = x^qq % N
+            c = c + ZZ(1)
+            if c >= ZZ(20)
+                g = gcd(x - ZZ(1), N)
+                if g == ZZ(1)
+                    c = ZZ(0)
+                    j = i
+                    y = x
+                else
+                    i = j
+                    x = y
+                    w = 1
+                end
+            end
+        end
+    end
+    w = 1
+
+    while w == 1
```

Comment:
Huh, this loop never changes, `w`, does it? So it might as well do `while true` ?

Or perhaps `while i <= k` and then move the code `if i > k  return ...` below after this loop?